client/concordclient/include/client/concordclient/concord_client.hpp

```diff
@@ -130,14 +132,59 @@ class ConcordClient {
 
   // Register a callback that gets invoked for every validated event received.
   // void callback(EventGroup);
+  // Return subscriber ID used to unsubscribe.
   template <class CallbackT>
   void subscribe(const SubscribeRequest& request,
                  const std::unique_ptr<opentracing::Span>& parent_span,
                  CallbackT callback);
+
+  // Unsubscribe with the subscriber ID retrieved from a prior call to subscribe.
+  // Note, if the caller doesn't unsubscribe and no runtime error occurs then resources
+  // will be occupied forever.
   void unsubscribe();
 
  private:
+  logging::Logger logger_;
   ConcordClientConfig config_;
+
+  // TODO: Allow multiple subscriptions
+  std::atomic_bool stop_subscriber_{true};
+  std::thread subscriber_;
 };
 
+template <class CallbackT>
+void ConcordClient::subscribe(const SubscribeRequest& request,
+                              const std::unique_ptr<opentracing::Span>& parent_span,
+                              CallbackT callback) {
+  if (not stop_subscriber_) {
+    LOG_ERROR(logger_, "subscription already in progress - unsubscribe first");
+    return;
+  }
+
+  stop_subscriber_ = false;
+  std::atomic_bool* stop = &stop_subscriber_;
+  subscriber_ = std::thread([stop, request, callback] {
+    while (not(*stop)) {
+      // Note: The following returns an artificial event group.
+      // This will be replaced with the actual thin replica client integration.
+      // The thread is in place to simulate the asynchronous subscription.
+      EventGroup eg;
+      eg.id = request.event_group_id;
+      std::string event = std::to_string(request.event_group_id);
+      eg.events.push_back({event.begin(), event.end()});
+      std::chrono::duration time_now = std::chrono::system_clock::now().time_since_epoch();
+      eg.record_time = std::chrono::duration_cast<std::chrono::microseconds>(time_now);
+      eg.trace_context = {};
+
+      callback(eg);
```

Comment:
I am not aware of the size of `EventGroup` but I would suggest changing the signature of the `callback` to accept an rvalue reference to `EventGroup`. Especially, because it is never reused in the thread.

Reply:
Good catch - thanks!